### PR TITLE
Enable timeout `ensureSynchronized` everywhere (Solution #2)

### PR DESCRIPTION
**Pull request by curtisman**
Monkey patch Mocha's `it` function to wrap test functions and capture how long the test is supposed to run until timeout and patch the `resetTimeout` and `clearTimeout` function on the test to trap the timeout tracking in mocha.  It will set up a parallel timeout that will reject a promise for `timeoutPromise` right before the test would timeout.

Remove the 5 min timeout for `ensureSynchronized` (which is unless for test) and use the now default behavior of the test timeout captured above. The purpose is to provide a better error message and stack if `ensureSynchronized` got stuck.

Also, change the behavior for `timeoutPromise`:
- when duration is <=0 or Infinity. Instead of waiting for the default amount, it will now mean it won't timeout at all. This is so that we can avoid timeout when debugging. 
- By default, when no duration is specified, `timeoutPromise` has no affect.  Only if you are inside the test function that it will use the test timeout.

Fix [AB#2175](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/2175)

This is an alternative solution to PR #12278

### CODE REVIEW

**Comment by curtisman** [comment]
The changes the behavior for what @anthony-murphy introduced with the original `timeoutPromise` concept, but it will get use more reliable timeout based on the test's timeout setup.  What do you think?

Also is a little bit uglier, using monkey patching, but it gets the job done.  Let me know if you have any other ideas.

File: packages/test/test-utils/src/timeoutUtils.ts [hunk 1]
```diff
@@ -4,15 +4,120 @@
...
+        return this.promise;
+    }
+
+    public getTimeout() {
+        return this.timeout;
+    }
+
+    private constructor() {
+        this.promise
+            = new Promise((resolve, reject: PromiseRejectFunc) => { this.reject = reject; });
```
Comment (anthony-murphy) [inline packages/test/test-utils/src/timeoutUtils.ts:57]: this is pretty much the same thing as our differed promise. could just reuse that

File: packages/test/test-utils/src/timeoutUtils.ts [hunk 2]
```diff
@@ -4,15 +4,120 @@
...
+    public static initialize() {
+        TestTimeout.instance = new TestTimeout();
+    }
+    public static reset(runnable: Mocha.Runnable) {
+        TestTimeout.clear();
+        TestTimeout.instance.resetTimer(runnable);
+    }
+
+    public static clear() {
+        if (TestTimeout.instance.rejected) {
```
Comment (anthony-murphy) [inline packages/test/test-utils/src/timeoutUtils.ts:36]: instance will be undefined if not in mocha. surprised this compiles. we must have strict null of in the test project

File: packages/test/test-utils/src/timeoutUtils.ts [hunk 3]
```diff
@@ -4,15 +4,120 @@
...
+
+        // Check the test timeout setting
+        const timeout = runnable.timeout();
+        if (!(Number.isFinite(timeout) && timeout > 0)) { return; }
+
+        // subtract a buffer
+        this.timeout = Math.max(timeout - timeBuffer, 1);
+
+        // Set up timer to reject near the test timeout.
+        this.timer = setTimeout(() => {
```
Comment (anthony-murphy) [inline packages/test/test-utils/src/timeoutUtils.ts:74]: this feels like a lot of complexity to just create a timer. i think i'd prefer to just make timeoutPromise smarter, and pull the timeout off the test in the default case

Comment (anthony-murphy) [reply packages/test/test-utils/src/timeoutUtils.ts:74]: i'm hoping we could just do something like `timeout = runnable.duration - runnable.timeout() - timeBuffer`

Comment (anthony-murphy) [reply packages/test/test-utils/src/timeoutUtils.ts:74]: if not. the below code in `if (globalThis.getMochaModule !== undefined)` could just set some var to be the start time in  runnablePrototype.resetTimeout  which would take the role of runnable.duration above

File: packages/test/test-utils/src/timeoutUtils.ts (hunk 2 above)
Comment (curtisman) [reply packages/test/test-utils/src/timeoutUtils.ts:36]: strict null check is on, but it must be a hole in type script type detection.

File: packages/test/test-utils/src/timeoutUtils.ts (hunk 3 above)
Comment (curtisman) [reply packages/test/test-utils/src/timeoutUtils.ts:74]: `resetTimeout` get calls after the async test function returns the promise, it is too late to influence the time `timeoutPromise` uses

Comment (curtisman) [reply packages/test/test-utils/src/timeoutUtils.ts:71]: Also, capture the end time was my original solution.  But that doesn't work if the call to set the timeout is within the test itself.   So we will have reset/clear/reset/clear.   We need to have a way to cancel and change the time out for `timeoutPromise` that is already in flight.

File: packages/test/test-utils/src/timeoutUtils.ts (hunk 2 above)
Comment (anthony-murphy) [reply packages/test/test-utils/src/timeoutUtils.ts:34]: you should add ` | undefined` to the type declaration of `instance` and handle undefined

File: packages/test/test-utils/src/timeoutUtils.ts [hunk 4]
```diff
@@ -4,15 +4,117 @@
...
+    // eslint-disable-next-line @typescript-eslint/unbound-method
+    const oldClearTimeoutFunc = runnablePrototype.clearTimeout;
+    runnablePrototype.clearTimeout = function(this: Mocha.Runnable) {
+        TestTimeout.clear();
+        oldClearTimeoutFunc.call(this);
+    };
+}
+
 export interface TimeoutWithError {
+    // Timeout duration in milliseconds, if it is > 0 and not Infinity
```
Comment (anthony-murphy) [inline packages/test/test-utils/src/timeoutUtils.ts:107]: comments should be `/***` and use ts doc notation for docs and intellisense

File: packages/test/test-utils/src/timeoutUtils.ts [hunk 5]
```diff
@@ -61,3 +163,33 @@ export async function timeoutPromise<T = void>(
...
+// Create a promise based on test timeout and the timeout options
+export async function timeoutPromise<T = void>(
+    executor: (resolve: (value: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void,
+    timeoutOptions: TimeoutWithError | TimeoutWithValue<T> = {},
+): Promise<T> {
+    // create the timeout error outside the async task, so its callstack includes
+    // the original call site, this makes it easier to debug
+    const err = timeoutOptions.reject === false
+        ? undefined
+        : new Error(timeoutOptions.errorMsg ?? "Test timed out");
```
Comment (anthony-murphy) [inline packages/test/test-utils/src/timeoutUtils.ts:176]: should we pass this error into getTimeoutPromise rather than create a different one there?

Comment (anthony-murphy) [reply packages/test/test-utils/src/timeoutUtils.ts:176]: seems strange we'll get different failures with different stack depending on timeout path

File: packages/test/test-utils/src/timeoutUtils.ts (hunk 3 above)
Comment (curtisman) [reply packages/test/test-utils/src/timeoutUtils.ts:71]: No. mocha parallelism is achieved via different process.

File: packages/test/test-utils/src/timeoutUtils.ts (hunk 5 above)
Comment (curtisman) [reply packages/test/test-utils/src/timeoutUtils.ts:176]: don't we want to distinguish explicit timeout version "test" timeout?

**Review by anthony-murphy — APPROVED** [review]

File: packages/test/test-utils/src/timeoutUtils.ts (hunk 5 above)
Comment (anthony-murphy) [reply packages/test/test-utils/src/timeoutUtils.ts:176]: maybe. i assume with this we'll move away from duration based timeout in general

File: packages/test/test-utils/src/timeoutUtils.ts (hunk 2 above)
Comment (curtisman) [reply packages/test/test-utils/src/timeoutUtils.ts:34]: I changed it to be always initialized.

File: packages/test/test-utils/src/timeoutUtils.ts (hunk 4 above)
Comment (curtisman) [reply packages/test/test-utils/src/timeoutUtils.ts]: Done.

**Pull request merged**